cluster-autoscaler/simulator/delta_cluster_snapshot.go

```diff
@@ -0,0 +1,490 @@
+	if data == nil {
+		return nil
+	}
+	if data.nodeInfoList == nil {
+		data.nodeInfoList = data.buildNodeInfoList()
+	}
+	return data.nodeInfoList
+}
+
+// Contains costly copying throughout the struct chain. Use wisely.
+func (data *internalDeltaSnapshotData) buildNodeInfoList() []*schedulernodeinfo.NodeInfo {
+	baseList := data.baseData.getNodeInfoList()
+	totalLen := len(baseList) + len(data.nodeInfoMap)
+	var nodeInfoList []*schedulernodeinfo.NodeInfo
+
+	if len(data.deletedNodeInfos) > 0 {
+		nodeInfoList = make([]*schedulernodeinfo.NodeInfo, 0, totalLen+100)
+		for _, bni := range baseList {
+			if data.deletedNodeInfos[bni.Node().Name] {
+				continue
+			}
+			nodeInfoList = append(nodeInfoList, bni)
+		}
+	} else {
+		nodeInfoList = make([]*schedulernodeinfo.NodeInfo, len(baseList), totalLen+100)
+		copy(nodeInfoList, baseList)
+	}
+
+	for _, dni := range data.nodeInfoMap {
+		nodeInfoList = append(nodeInfoList, dni)
+	}
+
+	return nodeInfoList
+}
+
+// Convenience method to avoid writing loop for adding nodes.
+func (data *internalDeltaSnapshotData) addNodes(nodes []*apiv1.Node) error {
+	for _, node := range nodes {
+		if err := data.addNode(node); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) addNode(node *apiv1.Node) error {
+	nodeInfo := schedulernodeinfo.NewNodeInfo()
+	if err := nodeInfo.SetNode(node); err != nil {
+		return fmt.Errorf("cannot set node in NodeInfo: %v", err)
+	}
+	return data.addNodeInfo(nodeInfo)
+}
+
+func (data *internalDeltaSnapshotData) addNodeInfo(nodeInfo *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.nodeInfoMap[nodeInfo.Node().Name]; found {
+		return fmt.Errorf("node %s already in snapshot", nodeInfo.Node().Name)
+	}
+	data.nodeInfoMap[nodeInfo.Node().Name] = nodeInfo
+	if data.nodeInfoList != nil {
+		data.nodeInfoList = append(data.nodeInfoList, nodeInfo)
+	}
+	return nil
+}
+
+func (data *internalDeltaSnapshotData) clearCaches() {
+	data.nodeInfoList = nil
+	data.clearPodCaches()
+}
+
+func (data *internalDeltaSnapshotData) clearPodCaches() {
+	data.podList = nil
+	data.podListChunks = nil
+	data.totalPods = 0
+	data.havePodsWithAffinity = nil
+}
+
+func (data *internalDeltaSnapshotData) updateNode(node *schedulernodeinfo.NodeInfo) error {
+	if _, found := data.nodeInfoMap[node.Node().Name]; found {
+		data.removeNode(node.Node().Name)
+	}
+
+	return data.addNodeInfo(node)
+}
+
+func (data *internalDeltaSnapshotData) removeNode(nodeName string) error {
+	_, foundInDelta := data.nodeInfoMap[nodeName]
+	if foundInDelta {
+		// If node was added or modified within this delta, delete this change.
+		delete(data.nodeInfoMap, nodeName)
+	}
+
+	_, foundInBase := data.baseData.getNodeInfo(nodeName)
+	if foundInBase {
+		// If node was found in the underlying data, mark it as deleted in delta.
+		data.deletedNodeInfos[nodeName] = true
+	}
+
+	if !foundInBase && !foundInDelta {
+		// Node not found in the chain.
+		return errNodeNotFound
```

Comment:
I don't know. It feels like it should depend on our use of this method, but our current use of it is "never remove a node from snapshot", since scale down simulations don't yet use packing in snapshot. Can we resolve it when we migrate that part of the code?